src/main/java/com/opentable/kafka/util/OffsetMetrics.java

```diff
@@ -0,0 +1,183 @@
+            if (sizes.isEmpty()) {
+                badTopics.add(topic);
+                return;
+            }
+            for (final int part : sizes.keySet()) {
+                builder.put(String.format("%s.partition.%d.size",   topic, part), new AtomicLongGauge());
+                builder.put(String.format("%s.partition.%d.offset", topic, part), new AtomicLongGauge());
+                builder.put(String.format("%s.partition.%d.lag",    topic, part), new AtomicLongGauge());
+            }
+            builder.put(String.format("%s.total.size",   topic), new AtomicLongGauge());
+            builder.put(String.format("%s.total.offset", topic), new AtomicLongGauge());
+            builder.put(String.format("%s.total.lag",    topic), new AtomicLongGauge());
+        });
+
+        if (!badTopics.isEmpty()) {
+            throw new IllegalArgumentException(
+                    String.format("bad topic%s: '%s'", badTopics.size() != 1 ? "s" : "", badTopics));
+        }
+
+        metricMap = builder.build();
+
+        exec = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setNameFormat(String.format("%s-offset-metrics-%%d", groupId))
+                        .build()
+        );
+    }
+
+    @Override
+    public Map<String, Metric> getMetrics() {
+        return metricMap;
+    }
+
+    public void start() {
+        exec.scheduleAtFixedRate(this::poll, 0, pollPeriod.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    public void stop() {
+        try {
+            OTExecutors.shutdownAndAwaitTermination(exec, Duration.ofSeconds(5));
+        } catch (final InterruptedException e) {
+            LOG.error("worker thread shutdown interrupted", e);
+            Thread.currentThread().interrupt();
+        }
+        monitor.close();
+    }
+
+    @Override
+    public void close() {
+        stop();
+    }
+
+    private void poll() {
+        topics.forEach(topic -> {
+            try {
+                pollUnsafe(topic);
+            } catch (final Exception e) {
+                LOG.error("caught exception for topic {}", topic, e);
+            }
+        });
+    }
+
+    private void pollUnsafe(final String topic) {
+        final Map<Integer, Long> sizes = monitor.getTopicSizes(topic);
+        sizes.forEach((part, size) -> gauge(String.format("%s.partition.%d.size", topic, part)).set(size));
+        gauge(String.format("%s.total.size", topic)).set(sumValues(sizes));
+
+        Map<Integer, Long> offsets = monitor.getGroupOffsets(groupId, topic);
+        final Map<Integer, Long> lag;
+        if (offsets.isEmpty()) {
+            // Consumer may not be consuming this topic yet (or consumer might not exist).
+            // In case consumer existed previously, we set all offsets and lag to 0.
+            offsets = lag = sizes
+                    .keySet()
+                    .stream()
+                    .collect(
+                            Collectors.toMap(
```

Comment:
Thinking out loud, and I will mark if I see the answers
- Won't all this mess up with rebalancing (I guess you can listen but...)
- Should the entire collection be snapshottable or you will have race between each of your gauges.

Reply:
* A) The partition assignment might change but the partition would remain, so no issue. It feels like the interpretation might change upon rebalance depending on the offset saving strategy. Assuming durability it shouldn't be wrong.
* B) That would be nice but it's not an option.